### ISSUE TITLE
Possible to release a published artifact?

### ISSUE THREAD

**Issue by jeff-blaisdell** [hyleung/ratpack-zipkin#8]
Do you happen to have this being published anywhere?

I searched bintray and maven and don't see anything published yet.  We see use for this project internally at my organization, and we would probably be willing to contribute over time.

**Comment by hyleung** [comment]
@jeff-blaisdell at the moment, the work in this repo is very much a work in progress. 

The server span (SS, SR) stuff seems to work fine. At least it was the last time I looked at it - both Ratpack _and_ Brave have progressed a fair bit during the life of this little project so a lot of things can change as the versions of those dependencies bump.

The story on the client span (CS, CR) is a little less complete. The problem I had was that default Ratpack HTTP client doesn't currently have any support for interceptors (something I'm trying to get working here - https://github.com/ratpack/ratpack/issues/975). In its current state, the client span tracing definitely isn't there yet.

If you just need the server span tracing then you could probably just rip out the server pieces and leave the client stuff out. Come to think of it,  I might have a look at that - if only so that I have less WIP to juggle.

**Comment by jeff-blaisdell** [comment]
That would be fantastic.  We've already started trying to utilize this project based off an earlier version, and we're basically trying to figure out how best we can contribute and not be blocked by the current state of master / ratpack interceptors.

Curious, did you run into a specific problem with your early approach w/ ZipkinHttpClient essentially wrapping the Ratpack HttpClient?

If you had a branch the was compilable w/ a current version of ratpack we could contribute there, and merge back into master for features that make sense.  Then master could contain the latest stuff w/ dependent on Ratpack http interceptors.  Would that work for you?

**Comment by hyleung** [comment]
I think the thing with the ZipkinHttpClient was, to be honest, that it felt sort of hacky. I thought the Ratpack HttpClient could have done with proper interceptor support, since that would hopefully mean that the internals of Ratpack's httpclient could change and the Zipkin instrumentation would not break. Anyhow, I got side-tracked working on that. 

I'm not sure where that code is in terms of compatibility with the current Ratpack release. TBH, I haven't done a tonne of testing. The http client stuff definitely got a lot of churn recently on Ratpack master. 

I can create that branch as you suggested - that should be fairly easy.  And if you need an actual release (tho' probably at this point, a snapshot would make more sense), I can look into that. 

Btw, just for context - eventually the plan was to bring this module back into Ratpack proper (https://github.com/ratpack/ratpack/issues/913).

**Comment by hyleung** [comment]
@jeff-blaisdell btw, are you using Kafka or Scribe? I suppose either way we'd want that included - so I might have to do some cherry-picking.

**Comment by codefromthecrypt** [comment]
For transport side, might want to have a look at this, which is the base
library recently made for that..
https://github.com/openzipkin/zipkin-reporter-java

**Comment by hyleung** [comment]
@jeff-blaisdell think what I'm going to do is revert the last 4 commits (c9bc5b4 - 46f383d) on master. I didn't anticipate anybody other than me hacking on this so that stuff might have some dependencies on WIP for ratpack/ratpack#975 - and ~~I don't have a good idea of when that will get done~~ that definitely wouldn't be going into Ratpack until at least 1.5.0. I noticed you've forked this repo already - otherwise I'd probably just rebase master.

After that I'd probably try to bring the Ratpack and Zipkin dependencies up to date, then start looking maybe setting up the release automation.

**Comment by jeff-blaisdell** [comment]
> are you using Kafka or Scribe?

Actually, neither.  We've written a custom collector that pushes to SQS; which may be OSS'd in the future.

thanks @hyleung I'll update our fork and see if it works for us.

**Comment by hyleung** [comment]
@jeff-blaisdell ah, cool. Maybe Zipkin reporter would be interesting to look into then. Do you think you'd be able to sort out your fork if I ended up doing a rebase? If so, I'd prefer to go that route so that we keep master free of WIP.

**Comment by jeff-blaisdell** [comment]
Yeah not a problem.  Rebase away.

**Comment by hyleung** [comment]
@jeff-blaisdell alright, I've bumped both Ratpack and Brave and made some small changes to get things working with the most recent version of Ratpack. I can start looking into releasing a version next (tho' I'd like to add some more tests too). All the WIP that requires my Ratpack fork is on a separate branch now.

Btw, saw this (https://github.com/openzipkin/brave/issues/228) - looks like somebody else is working on an SQS reporter, too! 😄

**Comment by jeff-blaisdell** [comment]
very nice!  thanks for all your work @hyleung.  This project feels promising.

**Comment by hyleung** [comment]
np - glad somebody is finding it useful. Should be pushing snapshot builds off master branch CI now: https://oss.sonatype.org/content/repositories/snapshots/com/github/hyleung/ratpack-zipkin/

After I get some of those extra tests in place, I'll probably move the `example` module out since I really only use it to manually(!) test things out.

**Comment by hyleung** [comment]
K, I remember why I didn't have those tests in the first place - the HttpClient stuff is a bit tricky to unit test - the setups are kinda gnarly. I wanted to write a test fixture similar to the ones that are used for HttpClient in the main Ratpack project, but the base types for those tests are in "internal" modules (`ratpack-test-internal`) and I didn't want to reproduce all that. 

The more I looked into it, the more I thought that it would be better if Ratpack had proper http client request/response interceptors - hence the big detour to ratpack/ratpack#975.

Anyhow, I'm going to skip the tests and do that clean up (I will probably move the main code up to the root since `example` will be gone), do a bit of manual testing (bleh) and then think about an actual release - how does that sound, @jeff-blaisdell?

Big caveat tho' - I had to revisit my own damn GH comments to figure this out - I don't think the way that `ZipkinHttpClient` is implemented will work for stuff other than `GET` and `POST` (https://github.com/ratpack/ratpack/issues/913#issuecomment-215995398).

**Comment by hyleung** [comment]
Verified that only `GET` and `POST` will work with the current implementation - created a separate issue to track/discuss the workaround for that here (https://github.com/hyleung/ratpack-zipkin/issues/9)

**Comment by hyleung** [comment]
@jeff-blaisdell it looks like you guys are already doing your own releases off the SmartThings fork(?). I assume you don't need a proper public release off of this repo, at least for now? Cos then I'll probably close this issue and maybe open a separate one to track all the things that would go into a 1.0.0.

**Comment by hyleung** [comment]
Tracking any odds and ends for 1.0.0 on the milestone - https://github.com/hyleung/ratpack-zipkin/milestone/1
